### PR TITLE
Feature/generation interface

### PR DESCRIPTION
**Pull request by mschop**
Rough design for the data structure definition.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,21 +13,6 @@ pids
...
-.grunt
-
-# node-waf configuration
-.lock-wscript
-
-# Compiled binary addons (http://nodejs.org/api/addons.html)
-build/Release
-
 # Dependency directory
 node_modules
```
Comment (mbj) [inline .gitignore]: Example: this one is intend to be repo root relative but also covers the `/foo/node_modules` which is likely not intend.

**Comment by mbj** [comment]
@mschop Is this ready for review yet?

**Comment by mschop** [comment]
@mbj It's just a rough design of the api. I thought you could give me feedback, if I am on a wrong track somewhere. I'm aware that there is currently no implementation. I have no idea what the right moment is to ask for a code review. I have no experience in this area. Should I wait with that until I have working functionality?

**Comment by mbj** [comment]
> @mbj It's just a rough design of the api. I thought you could give me feedback, if I am on a wrong track somewhere. I'm aware that there is currently no implementation. I have no idea what the right moment is to ask for a code review. I have no experience in this area. Should I wait with that until I have working functionality?

Maybe you should start more "primitive", with only implementing a specific instance, not thinking about general abstractions.

As you said, lots of the code right now is "disconnected", it shows infrastructure that is effectively dead as there are no clients.

What about starting in a way to only represet and produce nested `<p>` tags, no doctype, no text nodes etc. Just something very primitive. Get this correct and minimal THAN add more tags, reduce duplication, than add text nodes, reduce duplication, than add self closing tags, reduce duplication and so on.

This will be a far more educating and iterative process than a big one off design that can unless it implements all features not be proven correct or wrong. Its simply not evaluatable.

**Comment by mbj** [comment]
@mschop So you merged a PR by yourself, this is an antipattern.

**Pull request merged**